libminifi/src/core/Processor.cpp

```diff
@@ -289,6 +303,109 @@ bool Processor::isWorkAvailable() {
   return hasWork;
 }
 
+// must hold the graphMutex
+void Processor::updateReachability(const std::lock_guard<std::mutex>& graph_lock, bool force) {
+  bool didChange = force;
+  for (auto& outIt : out_going_connections_) {
+    for (auto& outConn : outIt.second) {
+      auto connection = std::dynamic_pointer_cast<Connection>(outConn);
+      if (!connection) {
+        continue;
+      }
+      auto dest = std::dynamic_pointer_cast<const Processor>(connection->getDestination());
+      if (!dest) {
+        continue;
+      }
+      if (reachable_processors_[connection].insert(dest).second) {
+        didChange = true;
+      }
+      for (auto& reachedIt : dest->reachable_processors_) {
+        for (auto &reached_proc : reachedIt.second) {
```

Comment:
O(scary), but good enough in practice. :D

Reply:
I'm all for a O(better) solution 🙂